### PR TITLE
#1380 dependencies updating

### LINKED ISSUE #1380: Dependency updating
## Is your feature request related to a problem? Please describe.

Need to upgrade dependencies. Security problems and to stay up to date

This should be done once 3.0.0 is released

## Describe the solution you'd like

Upgrade dependencies

## Describe alternatives you've considered

N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #1380 

**This was a real annoying one. It should be merged AFTER 3.0.0 is released, I think! But please don't let it get behind :(**

#### Upgraded `react-native` and `react`
- `react-native`: https://github.com/facebook/react-native/releases
    - AndroidX migration: `react-native` is now fully using AndroidX. This means other native dependencies need to use it also - https://developer.android.com/jetpack/androidx - this means we can remove the forced downgrading of google play services. Not 100% what this means for our native BLE module, but I'll find out when I get to it. I don't think there will be changes
    - Auto linking - Native modules are now auto linked when the app is built/compiled. So I removed all the manual linking of our current native modules. This also means the way to integrate a native module (previously by exposing a singleton of a module in `MainApplication` has changed. It's done automatically now through `PackageList.getPackages()`
    - Facebook are pushing the use of Flipper https://fbflipper.com/ - I haven't enabled it, I can do in a new PR as I think it looks pretty sweet. It's also open to plugins, so I'd imagine it would get pretty badass
    - "Fast refresh" - enabled by default, it's pretty quick but can be annoying. Can keep it enabled for now and see how everyone likes it, I think!
    - `react-native` upped `react` version: Main thing here was that `refs` can't be used for functional components. The only ref we actually used was with the navigator. This ref wasn't needed at all and i've just removed it
    - Hermes support: Realm isn't supporting it yet, so have to wait to use it
  
#### `react-native-device-info`
- This was upped a lot. Mainly for androidX support. It also completely changed a lot of it's api. We only used one method that changed: `getUniqueID` -> `getUniqueId`
https://github.com/react-native-community/react-native-device-info/wiki/V2-to-V3-Migration-Guide
https://github.com/react-native-community/react-native-device-info/wiki/V3-to-V4-Migration-Guide

#### `react-redux` / `redux`
- `react-redux` has a bunch of hooks you can use over `connect`
- new `batch` api! this will only cause a single re-render!
```
batch(() => {
            dispatch(increment());
            dispatch(increment());
        })
```

#### `react-native-gesture-handler`, `react-native-localization`, `react-native-vector-icons`, `react-native-modalbox` , `react-native-fs`, `bugsnag-react-native`, `realm`
- These were all just bug fixes / androidX changes

#### Babel/eslint/prettier
- bug fixes aaaand
- new syntax available
    - https://babeljs.io/docs/en/babel-plugin-proposal-nullish-coalescing-operator
    - https://babeljs.io/docs/en/babel-plugin-proposal-optional-chaining
    - https://babeljs.io/docs/en/babel-plugin-proposal-class-properties
    - https://github.com/tc39/proposal-private-methods

This also brings quite a few more warnings when running the app. I think we can address these in new PRs. For example use of depracated life cycle methods such as `componentWillUpdate`. These will be completely gone in `react 17` - so we should rename them to `UNSAFE_xx` so that they will continue to work.

## Testing

Steps to reproduce or otherwise test the changes of this PR:

- [ ] Can run `yarn build` and build apk's "as per normal" (@wlthomson)
- [ ] Can run `react-native run-android` and the app runs "as per normal"

I dunno what else to put 🤷‍♂ 

### Related areas to think about

- https://dependabot.com/?
- I think this fixes all the security warnings, mightv'e missed something but easiest to just check after this is merged 🤷‍♂ 
- I didn't upgrade `react-navigation` as it is a little more involved (https://github.com/react-navigation/react-navigation/releases?after=v4.0.4) and should be done in a new PR. I think it is important that gets done though

### CODE REVIEW

File: android/gradle.properties [hunk 1]
```diff
@@ -45,3 +45,5 @@ MYAPP_RELEASE_KEY_PASSWORD=sussolpikelets
 
 android.useAndroidX=true
 android.enableJetifier=true
+
+org.gradle.jvmargs=-XX\:MaxHeapSize\=1024m -Xmx1024m
```
Comment (wlthomson) [inline android/gradle.properties]: Related reference: https://scans.gradle.com/. Seems like a handy tool for debugging build issues.

File: android/app/build.gradle [hunk 2]
```diff
@@ -194,7 +194,6 @@ android {
 }
 
 dependencies {
-    implementation project(':realm')
```
Comment (wlthomson) [inline android/app/build.gradle]: Were these removals done using `react-native unlink`, or manually (just curious)?

File: android/keystores/debug.keystore.properties [hunk 3]
```diff
@@ -1,4 +0,0 @@
-key.store=debug.keystore
```
Comment (wlthomson) [inline android/keystores/debug.keystore.properties]: Not sure if this is related, but it looks like a debug key is needed for `react-native start`? I'm not able to build due to missing key `/mobile/android/app/debug.keystore` 😢 .

Comment (joshxg) [reply android/keystores/debug.keystore.properties]: can you paste the error output please?

Comment (wlthomson) [reply android/keystores/debug.keystore.properties]: Snippet of `react-native run-android` output:

```
Execution failed for task ':app:validateSigningDebug'.
> Keystore file '/Users/williamthomson/Projects/sussol/mobile/android/app/debug.keystore' not found for signing config 'debug'.
```

Comment (joshxg) [reply android/keystores/debug.keystore.properties]: I'm building fine :( If you don't mind would you please try adding this back to see if it makes a difference to you as a first step? Sorry to put that on you! I'm gonna look into keys now as I am unsure about them (so dont go deleting stuff you dont know about?)

File: android/app/build.gradle (hunk 2 above)
Comment (joshxg) [reply android/app/build.gradle]: `react-native unlink` .. just.. very slowly

File: yarn.lock [hunk 4]
```diff
@@ -805,51 +912,98 @@
...
-    slash "^2.0.0"
+    strip-ansi "^5.2.0"
+    sudo-prompt "^9.0.0"
+    wcwidth "^1.0.1"
     ws "^1.1.0"
-    xcode "^2.0.0"
-    xmldoc "^0.4.0"
 
-"@react-navigation/core@^3.0.3", "@react-navigation/core@~3.5.0":
+"@react-navigation/core@^3.5.1", "@react-navigation/core@~3.5.0":
```
[the remark this answers is not in the document]
Comment (joshxg) [reply yarn.lock]: looks like you're right, but some dependency wants it 🤷‍♂ I don't particularly wanna go messing with it

File: android/gradle.properties (hunk 1 above)
Comment (joshxg) [reply android/gradle.properties]: some developers who have a lot of cores need a larger heap size. Sorry if you don't need this?

File: yarn.lock [hunk 5]
```diff
@@ -608,13 +652,20 @@
...
 
-"@babel/runtime@^7.0.0", "@babel/runtime@^7.3.1", "@babel/runtime@^7.4.5":
+"@babel/runtime@^7.0.0", "@babel/runtime@^7.4.5":
   version "7.6.0"
   resolved "https://registry.yarnpkg.com/@babel/runtime/-/runtime-7.6.0.tgz#4fc1d642a9fd0299754e8b5de62c631cf5568205"
   integrity sha512-89eSBLJsxNxOERC0Op4vd+0Bqm6wRMqMbFtV3i0/fbaWw/mJ8Q3eBvgX0G4SyrOOLCtbu98HspF8o09MRT+KzQ==
   dependencies:
     regenerator-runtime "^0.13.2"
 
+"@babel/runtime@^7.5.5":
```
[the remark this answers is not in the document]
Comment (joshxg) [reply yarn.lock]: have deduplicated ba8966c35130f5882d6ce235dca9eefeab3078b2

File: yarn.lock [hunk 6]
```diff
@@ -29,6 +29,26 @@
     semver "^5.4.1"
     source-map "^0.5.0"
 
+"@babel/core@^7.6.4":
```
[the remark this answers is not in the document]
Comment (joshxg) [reply yarn.lock]: have deduplicated ba8966c35130f5882d6ce235dca9eefeab3078b2

File: package.json [hunk 7]
```diff
@@ -94,7 +93,7 @@
     "jest": "^24.8.0",
     "lint-staged": "^8.1.0",
     "metro-react-native-babel-preset": "^0.51.1",
-    "prettier": "^1.18.2",
+    "prettier": "1.18.2",
     "react-test-renderer": "^16.7.0"
   }
 }
```
[the remark this answers is not in the document]
Comment (joshxg) [reply package.json]: I try to upgrade independently or at least smaller upgrades. It occurs to infrequently and has been too hard to get pushed through regardless of size.

www.dependabot.com 👍

File: android/keystores/BUCK [hunk 8]
```diff
@@ -1,8 +0,0 @@
-keystore(
```
[the remark this answers is not in the document]
Comment (joshxg) [reply android/keystores/BUCK]: I think it was used in older versions of RN

File: android/app/src/main/java/com/msupplymobile/MainApplication.java [hunk 9]
```diff
@@ -30,20 +22,11 @@ public boolean getUseDeveloperSupport() {
...
-            new AsyncStoragePackage(),
-            new RNScreensPackage(),
-          new RNGestureHandlerPackage(),
-          new RealmReactPackage(),
-          new VectorIconsPackage(),
-          new ReactNativeLocalizationPackage(),
-          new RNFSPackage(),
-          BugsnagReactNative.getPackage()
-      );
+
```
[the remark this answers is not in the document]
Comment (joshxg) [reply android/app/src/main/java/com/msupplymobile/MainApplication.java]: 0bb65d9f3eeddd9326e34569c2dd1b542e68678c

File: android/build.gradle [hunk 10]
```diff
@@ -13,7 +13,7 @@ buildscript {
         jcenter()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.3.2'
+        classpath 'com.android.tools.build:gradle:3.4.2'
```
[the remark this answers is not in the document]
Comment (joshxg) [reply android/build.gradle]: gradle version, upped now

40f3e90a81e6d166a6e585284477b491bc945252

File: android/build.gradle [hunk 11]
```diff
@@ -27,13 +27,20 @@ allprojects {
...
+        jcenter()
+        maven { url 'https://jitpack.io' }
     }
 }
 
 
-task wrapper(type: Wrapper) {
-    gradleVersion = '4.10.1'
+wrapper {
+    gradleVersion = '5.1.1'
```
[the remark this answers is not in the document]
Comment (joshxg) [reply android/build.gradle]: Just got it working and forgot to keep incrementing!

cc87219e7101e60e3eb7e5d3999499b0c13dc90a

File: package.json [hunk 12]
```diff
@@ -94,7 +93,7 @@
     "jest": "^24.8.0",
     "lint-staged": "^8.1.0",
     "metro-react-native-babel-preset": "^0.51.1",
-    "prettier": "^1.18.2",
+    "prettier": "1.18.2",
```
[the remark this answers is not in the document]
Comment (joshxg) [reply package.json]: 18d06b44eccd52e40b19ea4a0a284a9cc4f70b90


> it's good practice to use ^ and just specify the exact version in the lock file?

I dunno what this means?

File: src/mSupplyMobileApp.js [hunk 13]
```diff
@@ -223,9 +222,6 @@ class MSupplyMobileAppContainer extends React.Component {
           RightComponent={finaliseItem ? this.renderFinaliseButton : this.renderSyncState}
         />
         <ReduxNavigator
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/mSupplyMobileApp.js]: We need to sort our navigation out, it's problematic in my opinion.

However, I think navigation should be in redux. I think it does mix view and business logic, but I think alternatives, that I can think of, are worse.

Take `Stocktake` creation flow: 
- Press the button
- Trigger a dispatch (currently to `usePageReducer`, but lets just say it's redux, it doesn't matter) to create a `Stocktake`
- ???

If we need to navigate without using redux, we need to have a `ref` to the `Navigation` component to call `navigate()` on. Where does this go? If it's in redux, it's a side effect and where do we keep the ref? In the store? Doesn't seem right to keep a `ref` to the DOM in the store, so does it get passed with any action? Some sort of middleware which injects the ref could be an option, actually but technically we're still using redux for this, just in a weird way.

So alternatively after the dispatch, you call `navigate()` on the `ref` within a view. But you have to do this asynchronously after the dispatch, so you now have async dispatches. 

Ok, so take creation of new records out of redux, it shouldn't really be there anyway as it's a side effect. Where does it go now? In the view? That's no good either :(

File: package.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (wlthomson) [reply package.json]: I believe the way it works (from my understanding) is that in the package file you specify the minimal version needed and in the lock file you specify the actual version (versions?) being used (e.g `prettier: ^1.18.2` in the package file, but `prettier@^1.18.2: version "1.18.2"` in the lock).

File: android/gradle.properties (hunk 1 above)
Comment (wlthomson) [reply android/gradle.properties]: Nah, wasn't complaining about it, was just saying that this tool seems useful for debugging and possibly finding a good heap size.

File: android/keystores/debug.keystore.properties (hunk 3 above)
Comment (wlthomson) [reply android/keystores/debug.keystore.properties]: Not sure of the cause, have reverted a few things and still seems to occur. Have fixed by building an arbitrary key and using that:

```
keytool -genkey -v -keystore debug.keystore -storepass android -alias androiddebugkey -keypass android -keyalg RSA -keysize 2048 -validity 10000
```

Will go ahead and resolve this.

File: package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (wlthomson) [reply package.json]: Dependabot looks pretty awesome.

**Comment by diegoevangelisti** [comment]
### Testing

 - [x] Can run yarn build and build apk's "as per normal" (@wlthomson)
 - [x] Can run react-native run-android and the app runs "as per normal"

**Success**. All working well. 
Apk's created successfully. Also after executing `react-native run-android` the app is running as expected. (same behaviour for command `yarn start`)

**Pull request merged**